include/bit_algorithm_details.hpp

```diff
@@ -132,6 +132,119 @@ ForwardIt word_shift_right(ForwardIt first,
+constexpr auto _all_ones() {
+    return -1;
+} 
+
+// returns a word consisting of all zero bits
+constexpr auto _all_zeros() {
+    return 0;
+}
+
+// checks that the passed iterator points to the first bit of a word
+template <class It>
+bool _is_aligned_lsb(bit_iterator<It> iter) {
+    return iter.position() == 0;
+}
+
+// checks that maybe_end is one position past the last bit of base
+template <class It>
+bool _is_one_past_last_bit(bit_iterator<It> maybe_end, It&& base) {
+    return maybe_end.position() == 0 
+      && std::next(std::forward<It>(base)) == maybe_end.base();
+}
+
+// checks that two bit iterators point to the same word
+template <class It>
+constexpr bool _in_same_word(bit_iterator<It> lhs, bit_iterator<It> rhs) {
+    return lhs.base() == rhs.base();
+}
+
+// simple alias for right shift
+template <class WordType>
+WordType _shift_towards_lsb(WordType word, std::size_t n) {
+    return word >> n; 
+} 
+
+// simple alias for left shift
+template <class WordType>
+WordType _shift_towards_msb(WordType word, std::size_t n) {
+    return word << n;
+}
+
+/* Used to read partial/full words and pad any missing digits. Will not
+ * read outside of the word pointed to by the first iterator (see case 4) 
+ *
+ * Case 0: 01011101
+ *        L       F
+ * Case 1: 01011101 -> padded with 0s -> 00001101
+ *            L   F
+ * Case 2: 01011101 -> padded with 1s -> 01011111 
+ *        L    F 
+ * Case 3: 01011101 -> padded with 0s -> 00011100
+ *           L  F
+ * Case 4: 01011101 11111111 -> treated as 01011101
+ *           F           L                L  F
+ *
+ * Note: word is read from [first, last), meaning the element pointed
+ * to by last is not included in the read. if first == last, behavior
+ * is undefined
+ */
+template <class It>
+typename bit_iterator<It>::word_type _padded_read(bit_iterator<It> first, 
+    bit_iterator<It> last, const bit::bit_value bv) {
+
+    using word_type = typename bit_iterator<It>::word_type;
+
+    constexpr std::size_t num_digits = binary_digits<word_type>::value;
+    const std::size_t first_position = first.position();
+    const std::size_t last_position = last.position();
+    const word_type read = *(first.base());
+    constexpr word_type all_ones = _all_ones();
+
+    word_type mask;
+
+    if (_is_aligned_lsb(first)) {
+        if (_in_same_word(first, last)) {
+            // Case 1
+            if (bv == bit0) {
+                mask = _shift_towards_lsb(all_ones, num_digits - last_position);
+                return read & mask;
+            } else {
+                mask = _shift_towards_msb(all_ones, last_position); 
+                return read | mask;
+            }
+        } else {
+            // Case 0
+            return read;
+        }
+    } else {
+        if (!_in_same_word(first, last)) {
+            // Case 2
+            if (bv == bit0) {
+                mask = _shift_towards_msb(all_ones, first_position); 
+                return read & mask;
```

Comment:
Why wouldn't we just shift return `_shift_towards_msb(read, first_position())`?

Reply:
I think there might be confusion about what the function is doing. If we have 01011101 with first pointed at the 3rd bit (0 indexed) and last pointed at one past the last bit i.e. ^01011^101 and we want to do a padded read by padding the bits outside the iterators with 0s, we should end up with 01011000. 01011101 << 3 = 11101000

Reply:
Yea so I  get that, but I guess maybe I should've made the same point as I did in the comment below. 

If we don't care about the position of the bit, why cant we just shift the original word instead? It would save us from having to perform the mask operation.

Reply:
I guess because then you can run into the issue of arbitrarily adding 0's or 1's to your word where it's not appropriate.

Reply:
yes but on this specific line you know that you are filling with 0s, so the only behavioral difference is the position. I suppose in `max`, you need to know the position. So you would have to increment the result of `tzcnt` with `first.position()` which is probably more expensive than the mask. 

I guess my point is we should only use `_padded_read` when we need to know the position of the bits as well. Maybe we could have an overloaded implementation of this such that we shift off the out of range bits instead of mask them? Again this only matters in the `bv=0` case, but it is still worth saving an operation IMO.

Reply:
Actually, it looks like [there is already arithmetic anyways](https://github.com/vreverdy/bit-algorithms/pull/14/files/c43a8e751697b24f13bc1bda17874460f51b7004#diff-7c6afd9d6651f72f61b9032e6f7569e0R39):

39: `return cursor + (first_set_bit_position - cursor.position());`

If instead we simply shifted off the out of range bits, we would then have
39: `return cursor + first_set_bit_position;`


And for [the case in the final word](https://github.com/vreverdy/bit-algorithms/pull/14/files/c43a8e751697b24f13bc1bda17874460f51b7004#diff-7c6afd9d6651f72f61b9032e6f7569e0R69), it makes more sense to mask the end off to keep the position the same, otherwise we end up with more arithmetic.

return `cursor + (last_set_bit_position - cursor.position());`

would end up being 

return `cursor + last_set_bit_position - (digits - last.position())`